pkg/cli/image/info/info.go

```diff
@@ -435,12 +404,67 @@ func (o *ImageRetriever) Run() error {
 						Config:        imageConfig,
 						Layers:        layers,
 						Manifest:      srcManifest,
-					}, manifestErr); err != nil {
+					}
+					if err = callbackFn(name, imageRetrieved, manifestErr); err != nil {
 						return err
 					}
 				}
 				return nil
 			})
 		}
 	})
+	if err != nil {
+		return nil, err
+	}
+	return imageRetrieved, nil
+}
+
+// ImageRetrieverOrList returns ImageRetriever and with the ManifestListCallbackFn provides a message
+// to distinguish errors between multiple manifests found and no manifests found, and lists os/arch options for manifestList upon error
+func (o *InfoOptions) ImageRetrieverOrList(src, filterByOS string) (*ImageRetriever, error) {
+	srcRef, err := imagesource.ParseReference(src)
+	if err != nil {
+		return &ImageRetriever{}, err
+	}
+	retriever := &ImageRetriever{
+		Image: map[string]imagesource.TypedImageReference{
+			src: srcRef,
+		},
+		FileDir:         o.FileDir,
+		SecurityOptions: o.SecurityOptions,
+		ManifestListCallback: func(from string, list *manifestlist.DeserializedManifestList, all map[digest.Digest]distribution.Manifest) (map[digest.Digest]distribution.Manifest, error) {
+			filtered := make(map[digest.Digest]distribution.Manifest)
+			for _, manifest := range list.Manifests {
+				if !o.FilterOptions.Include(&manifest, len(list.Manifests) > 1) {
+					klog.V(5).Infof("Skipping image for %#v from %s", manifest.Platform, from)
+					continue
+				}
+				filtered[manifest.Digest] = all[manifest.Digest]
+			}
+			if len(filtered) == 1 {
+				return filtered, nil
+			}
+			msg := fmt.Sprintf("no manifests retrieved that match filter %s", filterByOS)
```

Comment:
You should be getting filterByOS from o.FilterOptions here, not an argument

Reply:
duh, right, fixed.